### PR TITLE
optimize pathfinding

### LINKED ISSUE #564: Improve the pathfinding
- [x] Make `osrd-api` railjson endpoint return the geometry of tracksections (even if `includeGeom` is `false`)
- [x] Parse the geometry in `osrd-core`
- [ ] Adapt pathfinding endpoint to avoid middle ware computation

### PR DESCRIPTION
**Pull request by Tguisnet**
close #564

### CODE REVIEW

File: core/src/main/java/fr/sncf/osrd/api/PathfindingResult.java [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+import java.util.Collections;
+import java.util.List;
+
+public class PathfindingResult {
+    @Json(name = "route_paths")
+    public final List<RoutePathResult> routePaths = new ArrayList<>();
+    @Json(name = "path_waypoints")
+    public final List<PathWaypointResult> pathWaypoints = new ArrayList<>();
+
+    // make them private final
```
Comment (flomonster) [inline core/src/main/java/fr/sncf/osrd/api/PathfindingResult.java]: ?

File: core/src/main/java/fr/sncf/osrd/api/PathfindingResult.java [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+
+    /**
+     * pathfindingResultMake is used to create a class PathfindingResult with a good format to be sent to the middleware
+     * during this function, we add geometric information in the PathfindingResult
+     * @param finalPathsToGoal contains the routes that are use after the pathfinding
+     * @param infra current infra
+     * @return A pathfindingResult object ready to be sent to the middleware
+     */
+    @SuppressFBWarnings({"BC_UNCONFIRMED_CAST"})
+    public static PathfindingResult pathfindingResultMake(
```
Comment (flomonster) [inline core/src/main/java/fr/sncf/osrd/api/PathfindingResult.java]:
```suggestion
    public static PathfindingResult make(
```

File: core/src/main/java/fr/sncf/osrd/api/PathfindingResult.java [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+    public LineString geographic = null;
+    public LineString schematic = null;
+
+    /**
+     * pathfindingResultMake is used to create a class PathfindingResult with a good format to be sent to the middleware
+     * during this function, we add geometric information in the PathfindingResult
+     * @param finalPathsToGoal contains the routes that are use after the pathfinding
+     * @param infra current infra
+     * @return A pathfindingResult object ready to be sent to the middleware
+     */
```
Comment (flomonster) [inline core/src/main/java/fr/sncf/osrd/api/PathfindingResult.java]: We have as a principle to avoid making docs that paraphrase the code.
The `@return` and `@param infra` does not seem usefull.

File: core/src/main/java/fr/sncf/osrd/api/PathfindingResult.java [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+                res.add(route, trackSections);
+                if (j == routes.size() - 1) {
+                    // Add the given destination location to the steps output
+                    var lastTrack = trackSections.get(trackSections.size() - 1);
+                    var newStep = new PathWaypointResult(lastTrack.edge,
+                            lastTrack.getEndPosition());
+                    res.addStep(newStep);
+                }
+            }
+        }
```
Comment (flomonster) [inline core/src/main/java/fr/sncf/osrd/api/PathfindingResult.java:93]: This could be put in a function like:
```java
private static void fillRoutePath(
    ArrayDeque<FullPathArray<Route, BasicPathNode<Route>>> finalPathsToGoal,
    PathfindingResult pathfindingResult
) {
   // Copy paste
}
```

File: core/src/main/java/fr/sncf/osrd/api/PathfindingResult.java [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+        res.addGeometry(infra);
+        return res;
+    }
+
+    @SuppressFBWarnings({"BC_UNCONFIRMED_CAST"})
+    private static RouteLocation pathNodeToRouteLocation(BasicPathNode<Route> node) {
+        return new RouteLocation(node.edge, node.position);
+    }
+
+    void add(Route route, List<TrackSectionRange> trackSections) {
```
Comment (flomonster) [inline core/src/main/java/fr/sncf/osrd/api/PathfindingResult.java]:
```suggestion
    private void add(Route route, List<TrackSectionRange> trackSections) {
```

File: core/src/main/java/fr/sncf/osrd/utils/geom/Point.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || Point.class != o.getClass()) return false;
+        var point = (Point) o;
+        return Double.compare(point.x, x) == 0 && Double.compare(point.y, y) == 0;
+    }
+
+    @Override
```
Comment (flomonster) [inline core/src/main/java/fr/sncf/osrd/utils/geom/Point.java]: You can add the annotation `@ExcludeFromGeneratedCodeCoverage`

**Review by flomonster — CHANGES_REQUESTED** [review]

File: core/src/main/java/fr/sncf/osrd/utils/geom/LineString.java [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+            interval++;
+        }
+
+        var x1 = bufferX[interval];
+        var y1 = bufferY[interval];
+        var x2 = bufferX[interval + 1];
+        var y2 = bufferY[interval + 1];
+        var ratio = distance / lengths[interval];
+
+        return new Point(x1 + ratio * (x2 -x1),y1 + ratio * (y2- y1));
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/utils/geom/LineString.java]: missing space after comma

File: core/src/main/java/fr/sncf/osrd/utils/geom/LineString.java [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+    }
+
+    /**
+     * Interpolate a LineString
+     * @param distance normalize distance between 0 (origin) and 1 (endpoint)
+     * @return the point within the geometry at the given distance
+     */
+    public Point interpolateNormalized(double distance) {
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/utils/geom/LineString.java:171]: it's not a distance, it's a ratio

File: core/src/main/java/fr/sncf/osrd/utils/geom/Point.java [hunk 9]
```diff
@@ -0,0 +1,88 @@
+package fr.sncf.osrd.utils.geom;
+
+import com.squareup.moshi.*;
+
+import java.io.IOException;
+import java.util.Objects;
+
+public final class Point {
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/utils/geom/Point.java]: you can use a record instead of a class, which removes the need for implementing equals and hashcode

File: core/src/test/java/fr/sncf/osrd/utils/geom/LineStringTest.java [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    public void testConcatenate() {
+        var list = new ArrayList<LineString>();
+        list.add(LineString.make(new double[]{1.0, 4.0}, new double[]{2.0, 4.0}));
+        list.add(LineString.make(new double[]{4.0, 3.0}, new double[]{4.0, 3.0}));
+        list.add(LineString.make(new double[]{3.0, 4.0, 3.0}, new double[]{1.0, 4.0, 3.0}));
+
+        var expectedLineString = LineString.make(new double[]{1.0, 4.0, 3.0, 3.0, 4.0, 3.0}, new double[]{2.0, 4.0, 3.0, 1.0, 4.0, 3.0});
+        assertEquals(expectedLineString.getPoints(), LineString.concatenate(list).getPoints());
+    }
+}
```
Comment (multun) [inline core/src/test/java/fr/sncf/osrd/utils/geom/LineStringTest.java]: missing line feed

File: core/src/test/java/fr/sncf/osrd/utils/geom/PointTest.java [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+import java.io.IOException;
+
+class PointTest {
+    @Test
+    public void testDeserialize() throws IOException {
+        var geoJson = "{\"type\": \"Point\", \"coordinates\": [6.328125, 46.619261036171515]}";
+        var adapter = new Point.Adapter();
+        var point = adapter.fromJson(geoJson);
+    }
+}
```
Comment (multun) [inline core/src/test/java/fr/sncf/osrd/utils/geom/PointTest.java]: missing line feed

File: core/src/main/java/fr/sncf/osrd/railjson/schema/infra/RJSTrackSection.java [hunk 12]
```diff
@@ -16,8 +19,12 @@
 
     public List<RJSSlope> slopes;
     public List<RJSCurve> curves;
+    /** List of speed sections on the track section */
+    @Json(name = "speed_sections")
+    public List<RJSSpeedSection> speedSections;
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/railjson/schema/infra/RJSTrackSection.java:24]: this change is unrelated to your commit

File: core/src/main/java/fr/sncf/osrd/utils/geom/LineString.java [hunk 13]
```diff
@@ -146,7 +146,7 @@ private Point interpolate(double distance) {
         assert distance >= 0.;
         assert distance <= cumulativeLength;
         int interval = 0;
-        while (distance > lengths[interval]) {
+        while (interval < lengths.length - 1 && distance > lengths[interval]) {
             distance -= lengths[interval];
             interval++;
         }
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/utils/geom/LineString.java]: not a big deal, but you could create a cumuativeLengths array and run a binary search (see Arrays.binarySearch)

File: core/src/main/java/fr/sncf/osrd/utils/geom/LineString.java [hunk 14]
```diff
@@ -171,6 +171,51 @@ public Point interpolateNormalized(double distance) {
...
+        var newBufferY = new DoubleArrayList();
+
+        var firstPoint = interpolateNormalized(begin);
+        newBufferX.add(firstPoint.x);
+        newBufferY.add(firstPoint.y);
+
+        var intervalIndex = 1;
+        var currentLength = lengths[0];
+        // skip first points
+        while (begin >= currentLength / cumulativeLength && intervalIndex < lengths.length) {
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/utils/geom/LineString.java]: how about computing the absolute begin / end positions outside of the loop? this would remove these divisions and hopefully make things a tiny bit clearer

File: api/osrd_infra/views/pathfinding.py [hunk 15]
```diff
@@ -196,17 +197,17 @@ def compute_path(path, request_data, owner):
...
     waypoints, step_durations = parse_steps_input(request_data["steps"], infra)
     payload = request_pathfinding({"infra": infra.pk, "waypoints": waypoints})
+    path.geographic = json.dumps(payload.pop("geo_geom"))
+    path.schematic = json.dumps(payload.pop("sch_geom"))
 
     # Post treatment
     track_map = fetch_track_sections_from_payload(infra, payload)
     payload = compute_path_payload(infra, payload, step_durations, track_map)
-    geographic, schematic = get_geojson_path(payload, track_map)
+    # geographic, schematic = get_geojson_path(payload, track_map)
```
Comment (multun) [inline api/osrd_infra/views/pathfinding.py]: can you just remove the code?

File: core/src/main/java/fr/sncf/osrd/api/PathfindingResult.java [hunk 16]
```diff
@@ -0,0 +1,256 @@
...
+    public LineString schematic;
+
+    private PathfindingResult(ArrayList<RoutePathResult> routePaths, ArrayList<PathWaypointResult> pathWaypoints, LineString geographic, LineString schematic) {
+        this.routePaths = routePaths;
+        this.pathWaypoints = pathWaypoints;
+        this.geographic = geographic;
+        this.schematic = schematic;
+    }
+
+    public static PathfindingResult PathfindingResultMake(ArrayDeque<FullPathArray<Route, BasicPathNode<Route>>> finalPathsToGoal, Infra infra) throws InvalidInfraException {
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/api/PathfindingResult.java]: that's a weird name, can you just call it make? this way people would use PathfindingResult.make

**Review by flomonster — DISMISSED** [review]

File: core/src/main/java/fr/sncf/osrd/api/PathfindingResult.java [hunk 17]
```diff
@@ -0,0 +1,256 @@
...
+                            lastTrack.getEndPosition());
+                    // TODO
+                    //res.addStep(newStep);
+                }
+            }
+        }
+
+        ArrayList<RoutePathResult>routePath = null;
+        var geometry = addGeometry(infra, routePath);
+        return new PathfindingResult(null, null, geometry.get(0), geometry.get(1));
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/api/PathfindingResult.java]: what does the above code do? It seems to me like whatever it does isn't stored anywhere, and what's returned does not use it

File: core/src/main/java/fr/sncf/osrd/api/PathfindingRoutesEndpoint.java [hunk 18]
```diff
@@ -245,137 +202,4 @@ public Response act(Request req) throws IOException {
...
-            public void merge(PathWaypointResult other) {
-                suggestion &= other.suggestion;
-                if (!other.suggestion)
-                    return;
-                position = other.position;
-                id = other.id;
-            }
-        }
-    }
 }
```
Comment (flomonster) [inline core/src/main/java/fr/sncf/osrd/api/PathfindingRoutesEndpoint.java]: line feed missing

File: core/src/main/java/fr/sncf/osrd/utils/geom/LineString.java [hunk 19]
```diff
@@ -0,0 +1,306 @@
...
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class LineString {
+
+    /**
+     * A list of N coordinates (X component)
+     */
```
Comment (flomonster) [inline core/src/main/java/fr/sncf/osrd/utils/geom/LineString.java]:
```suggestion
    /** A list of N coordinates (X component) */
```

File: core/src/main/java/fr/sncf/osrd/utils/geom/LineString.java [hunk 20]
```diff
@@ -0,0 +1,306 @@
...
+
+                if (distance < 1e-5) {
+                    newBufferX.remove(newBufferX.size() - 1);
+                    newBufferY.remove(newBufferY.size() - 1);
+                } else {
+                    newCumulativeLengths.add(distance + newCumulativeLengths.get(newCumulativeLengths.size() - 1));
+                }
+            }
+            newBufferX.add(lineString.bufferX);
+            newBufferY.add(lineString.bufferY);
```
Comment (flomonster) [inline core/src/main/java/fr/sncf/osrd/utils/geom/LineString.java:125]:
```java
double lastCumulativeLength = 0;
if (!newCumulativeLengths.isEmpty())
    lastCumulativeLength = newCumulativeLengths[newCumulativeLengths.length - 1];
for (var cumLength : lineString.cumulativeLengths)
    newCumulativeLengths.add(cumLength + lastCumulativeLength);
```

File: core/src/main/java/fr/sncf/osrd/utils/geom/LineString.java [hunk 21]
```diff
@@ -0,0 +1,306 @@
...
+     * @param distance a distance between 0 and cumulativeLength
+     * @return the point within the geometry at the given distance
+     */
+    private Point interpolate(double distance) {
+        assert distance >= 0.;
+        assert distance <= cumulativeLengths[cumulativeLengths.length - 1];
+        var interval = Arrays.binarySearch(cumulativeLengths, distance);
+
+        if (interval < 0)
+            interval = Math.abs(interval) - 1;
```
Comment (flomonster) [inline core/src/main/java/fr/sncf/osrd/utils/geom/LineString.java]:
```suggestion
            interval = -interval - 1;
```

**Review by flomonster — CHANGES_REQUESTED** [review]

File: core/src/main/java/fr/sncf/osrd/utils/geom/LineString.java [hunk 22]
```diff
@@ -0,0 +1,306 @@
...
+        newBufferY.add(firstPoint.y());
+
+        var intervalBegin = Arrays.binarySearch(
+                cumulativeLengths,
+                begin * cumulativeLengths[cumulativeLengths.length - 1]
+        );
+        if (intervalBegin >= 0)
+            intervalBegin += 2;
+        else
+            intervalBegin = Math.abs(intervalBegin);
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/utils/geom/LineString.java]: I understand this because i've used binarySearch recently, but it's a pretty steep step, can you add a comment explaining why it works?

**Review by multun — DISMISSED** [review]

**Review by flomonster — APPROVED** [review]
